HHH-17507 HHH-17574 Fixed wrap() & unwrap() & updated unit tests

H-Lo opened the pull request:
- removed util class that was used for WrapperOptions
- fixed unwrap() metod in CurrencyJavaType
- unwrap() was already ok in YearJavaType
- fixed existing tests to avoid usage of (removed) util class in both CurrencyMappingTests and YearMappingTests
- added few new tests to test unwrap() method in both CurrencyMappingTests and YearMappingTests

<!-- Hibernate GitHub Bot issue links start -->
<!-- THIS SECTION IS AUTOMATICALLY GENERATED, ANY MANUAL CHANGES WILL BE LOST -->
https://hibernate.atlassian.net/browse/HHH-17507
https://hibernate.atlassian.net/browse/HHH-17574
<!-- Hibernate GitHub Bot issue links end -->

mbladel commented on hibernate-core/src/main/java/org/hibernate/type/descriptor/java/CurrencyJavaType.java:
```diff
@@ -35,21 +35,27 @@ public Currency fromString(CharSequence string) {
 
 	@SuppressWarnings("unchecked")
-	public <X> X unwrap(Currency value, Class<X> type, WrapperOptions options) {
+	public <X> X unwrap(final Currency value, final Class<X> type, final WrapperOptions options) {
```
No need to change the parameters to `final`.

mbladel commented on hibernate-core/src/main/java/org/hibernate/type/descriptor/java/CurrencyJavaType.java:
```diff
@@ -35,21 +35,27 @@ public Currency fromString(CharSequence string) {
 		if ( value == null ) {
 			return null;
 		}
+		if ( type.isInstance( value ) ) {
+			return (X) value;
+		}
```
```suggestion
		if ( Currency.class.isAssignableFrom( type ) ) {
			return (X) value;
		}
```

mbladel commented on hibernate-core/src/main/java/org/hibernate/type/descriptor/java/CurrencyJavaType.java:
```diff
@@ -35,21 +35,27 @@ public Currency fromString(CharSequence string) {
 
 	@Override
-	public <X> Currency wrap(X value, WrapperOptions options) {
+	public <X> Currency wrap(final X value, final WrapperOptions options) {
```
No need to change the parameters to `final`.

mbladel commented on hibernate-core/src/main/java/org/hibernate/type/descriptor/java/YearJavaType.java:
```diff
@@ -42,7 +42,7 @@ public Year fromString(CharSequence string) {
 	@SuppressWarnings("unchecked")
 	@Override
-	public <X> X unwrap(Year value, Class<X> type, WrapperOptions options) {
+	public <X> X unwrap(final Year value, final Class<X> type, final WrapperOptions options) {
```
No need to change the parameters to `final`.

mbladel commented on hibernate-core/src/main/java/org/hibernate/type/descriptor/java/YearJavaType.java:
```diff
@@ -67,11 +67,15 @@ public <X> X unwrap(Year value, Class<X> type, WrapperOptions options) {
 
 	@Override
-	public <X> Year wrap(X value, WrapperOptions options) {
+	public <X> Year wrap(final X value, final WrapperOptions options) {
```
No need to change the parameters to `final`.

mbladel reviewed (commented):
Thanks @H-Lo. This is almost perfect, just a few more suggestions. Also please squash everything in a single commit with both Jira issues at the start of the message.

Feel free to force-push to this PR instead of closing it and opening a new one.

H-Lo commented:
Sorry for the mess! I did force push and compare shows that finals are removed, but I'm not sure what happened with `isAssignableFrom` in `CurrencyJavaType`. I guess it's fixed too, but please do one more check.
That code was added to be the same as it was already in YearJavaType, So if you could check this too, it would be great.

mbladel commented:
@H-Lo the change is fine. Could you please squash all commits to a single one with a more meaningful message? Thanks.

mbladel reviewed (approved):
LGTM

H-Lo commented:
@mbladel Hi! Done squashing and removed some finals I added automatically without thiniking. Hope now you have minimal code changes to review comparing to master.

H-Lo commented:
Oh my! ou're going to have nightmares because of me.
After removing code 'if ( type.isInstance( value ) ) { return (X) value; }' unit test had to be updated.

Pushed again after running all tests.
`Tests passed: 16,512; Ignored 1,802 of 18314 tests - 6 min 42 sec `
